google/cloud/storage/internal/curl_client.cc

```diff
@@ -59,11 +76,21 @@ CurlClient::CurlClient(ClientOptions options)
     : options_(std::move(options)),
       share_(curl_share_init(), &curl_share_cleanup),
       storage_factory_(CreateHandleFactory(options_)),
-      upload_factory_(CreateHandleFactory(options_)) {
+      upload_factory_(CreateHandleFactory(options_)),
+      xml_upload_factory_(CreateHandleFactory(options_)),
+      xml_download_factory_(CreateHandleFactory(options_)) {
   storage_endpoint_ = options_.endpoint() + "/storage/" + options_.version();
   upload_endpoint_ =
       options_.endpoint() + "/upload/storage/" + options_.version();
 
+  if (std::getenv("CLOUD_STORAGE_TESTBENCH_ENDPOINT") != nullptr) {
+    xml_upload_endpoint_ = options_.endpoint() + "/xmlapi";
+    xml_download_endpoint_ = options_.endpoint() + "/xmlapi";
+  } else {
+    xml_upload_endpoint_ = "https://upload-storage.googleapis.com";
```

Comment:
Where do these come from?  I don't see these mentioned anywhere in the docs -- everything I can find just uses `<bucketname>.storage.googleapis.com/<objectname>`, or `storage.googleapis.com/<bucketname>/<objectname>`.

Reply:
From here:

https://cloud.google.com/storage/docs/request-endpoints

under "XML API". Took me a while to find them too (yuck), and the documentation claims better performance (maybe) with them:

> Optionally for uploads and downloads, Cloud Storage offers dedicated XML API endpoints. These upload and download endpoints may provide a small performance benefit when transferring large amounts of data.

Reply:
Oh! I see why I couldn't find them -- here, you've got "upload-storage" and "download-storage", instead of "storage-upload" and "storage-download".  I was searching for the reversed version :)

Reply:
Ooops, my bad. Fixed.